packages/shell-api/src/field-level-encryption.ts

```diff
@@ -0,0 +1,165 @@
+import { classPlatforms, hasAsyncChild, returnsPromise, ShellApiClass, shellApiClassDefault } from './decorators';
+import { Document, ReplPlatform, BinaryType } from '@mongosh/service-provider-core';
+import Collection from './collection';
+import Cursor from './cursor';
+import { DeleteResult } from './result';
+import { assertArgsDefined, assertArgsType } from './helpers';
+import { MongoshInvalidInputError } from '@mongosh/errors';
+
+/** Configuration options for using 'aws' as your KMS provider */
+declare interface awsKms {
+  aws: {
+    /** The access key used for the AWS KMS provider */
+    accessKeyId?: string;
+    /** The secret access key used for the AWS KMS provider */
+    secretAccessKey?: string;
+  };
+}
+
+/** Configuration options for using 'local' as your KMS provider */
+declare interface localKms {
+  local: {
+    /** The master key used to encrypt/decrypt data keys. A 96-byte long Buffer. */
+    key: BinaryType;
+  };
+}
+
+export interface ClientSideFieldLevelEncryptionOptions {
+  keyVaultClient?: any, /** This should be a Mongo class instance, but to avoid circular ref use any **/
+  keyVaultNamespace: string,
+  kmsProvider: awsKms | localKms,
+  schemaMap?: Document,
+  bypassAutoEncryption?: boolean;
+}
+
+@shellApiClassDefault
+@hasAsyncChild
+@classPlatforms([ ReplPlatform.CLI ] )
+export class ClientEncryption extends ShellApiClass {
+  public _mongo: any; // Mongo but any to avoid circular ref
+  public _libmongocrypt: any;
+
+  constructor(mongo: any) {
+    super();
+    this._mongo = mongo;
+    this._libmongocrypt = new mongo._serviceProvider.fle.ClientEncryption(
+      mongo._serviceProvider.getRawClient(),
+      {
+        options: this._mongo._fleOptions
+      }
+    );
+  }
+
+  @returnsPromise
+  encrypt(
+    encryptionId: BinaryType,
+    value: any,
+    encryptionAlgorithm: string
+  ): BinaryType {
+    assertArgsDefined(encryptionId, value, encryptionAlgorithm);
+    return this._libmongocrypt.encrypt(
+      value,
+      {
+        keyId: encryptionId,
+        algorithm: encryptionAlgorithm
+      }
+    );
+  }
+
+  @returnsPromise
+  decrypt(
+    encryptedValue: any
+  ): BinaryType {
+    assertArgsDefined(encryptedValue);
+    return this._libmongocrypt.decrypt(encryptedValue);
+  }
+}
+
+@shellApiClassDefault
+@hasAsyncChild
+@classPlatforms([ ReplPlatform.CLI ] )
+export class KeyVault extends ShellApiClass {
+  public _mongo: any; // Mongo but any to avoid circular ref
+  public _clientEncryption: ClientEncryption;
+  private _keyColl: Collection;
+  constructor(mongo: any, clientEncryption: ClientEncryption) {
+    super();
+    this._mongo = mongo;
+    this._clientEncryption = clientEncryption;
+    const [ db, coll ] = mongo._fleOptions.keyVaultNamespace.split('.');
+    if (coll.includes('.')) {
+      throw new MongoshInvalidInputError('keyVaultNamespace must be <db>.<coll>');
```

Comment:
This error is never going to be thrown, I think

Reply:
Ok can be removed

Reply:
I mean, I think, you probably want a condition like `if (!coll)` here instead – but a string split at `'.'` won’t contain `'.'` in the resulting parts

Reply:
Oops, the ts compiler complains "Object is possibly 'undefined'" so I think the error check should stay there? The namespace would get validated upon `MongoClient` connection but there's no way to indicate that to the compiler, from what I can tell?